zookeeper-api/src/main/java/org/apache/helix/zookeeper/zkclient/util/ZkPathRecursiveWatcherTrie.java

```diff
@@ -0,0 +1,180 @@
+import java.util.concurrent.locks.ReadWriteLock;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+import java.util.stream.Stream;
+
+import com.google.common.annotations.VisibleForTesting;
+import org.apache.helix.zookeeper.zkclient.RecursivePersistListener;
+
+
+/**
+ * ZkPathRecursiveWatcherTrie will be used for a registry for ZK persist recursive watcher.
+ * When persist recursive watcher is registered on path /X, ZK will send out data change for any
+ * data/child changing under the tree structure of /X. The event only include the path of changed
+ * ZNode.
+ * In ZkClient, when ever we get a dataChange event for path /X/Y/Z, we need to track back the path
+ * and notify all registered recursive persist listener and notify them about the node change.
+ * ref: https://zookeeper.apache.org/doc/r3.7.1/zookeeperProgrammers.html#sc_WatchPersistentRecursive
+ */
+public class ZkPathRecursiveWatcherTrie {
+
+  /** Root node of PathTrie */
+  private final TrieNode rootNode;
+
+  static class TrieNode {
+
+    final String value;   // Segmented ZNode path at current level
+    final Map<String, TrieNode> children; // A map of segmented ZNode path of next level to TrieNode
+    Set<RecursivePersistListener> recursiveListeners;
+    // A list of recursive persist watcher on the current path
+
+    /**
+     * Create a trie node with parent as parameter.
+     *
+     * @param value the value stored in this node
+     */
+    private TrieNode(String value) {
+      this.value = value;
+      this.children =
+          new HashMap<>(4);  // We keep the same number of init children side as Zk serer
+      this.recursiveListeners = new HashSet<>(4);
+    }
+
+
+    /**
+     * The value stored in this node.
+     *
+     * @return the value stored in this node
+     */
+    public String getValue() {
+      return this.value;
+    }
+
+    /**
+     * Add a child to the existing node.
+     *
+     * @param childName the string name of the child
+     * @param node the node that is the child
+     */
+    void addChild(String childName, TrieNode node) {
+      this.children.putIfAbsent(childName, node);
+    }
+
+    /**
+     * Return the child of a node mapping to the input child name.
+     *
+     * @param childName the name of the child
+     * @return the child of a node
+     */
+    @VisibleForTesting
+    TrieNode getChild(String childName) {
+      return this.children.get(childName);
+    }
+
+    /**
+     * Get the list of children of this trienode.
+     *
+     * @return A collection containing the node's children
+     */
+    @VisibleForTesting
+    Collection<String> getChildren() {
+      return children.keySet();
+    }
+
+    /**
+     * Get the set of RecursivePersistWatcherListener
+     * Returns an empty set if no listener is registered on the path
+     * @return
+     */
+    @VisibleForTesting
+    Set<RecursivePersistListener> getRecursiveListeners() {
+      return recursiveListeners;
+    }
```

Comment:
Open question on thread safety -- what's the plan for synchronization on READ? 
Should we consider a copy-on-write structure?

Reply:
TFTR. We will need to synchronization on read as well. I thinked about using other complex lock, after offline discussion, we decide to just use an simple `synchronization` keyword for simplicity. Since all event (read) are handled sequentially and persist recursive listener modification happens lot less frequent, we want to keep it simple for now.

Reply:
Had an offline discussion. We briefly discussed possibility of finer grain lock - concurrent hashmap for trie childMap and copyOnWrite for listeners. We both agree that having finer grain lock will help performance improvement. 
We decide to keep it simple for now and consider the opportunity for future performance improvement.